src/Testcontainers/Clients/TestcontainersClient.cs

```diff
@@ -259,8 +259,10 @@ public async Task<string> RunAsync(ITestcontainersConfiguration configuration, C
           .ConfigureAwait(false);
       }
 
-      if (!await this.images.ExistsWithNameAsync(configuration.Image.FullName, ct)
-        .ConfigureAwait(false))
+      var cachedImage = await this.images.ByNameAsync(configuration.Image.FullName, ct)
+        .ConfigureAwait(false);
+
+      if (configuration.ImagePullPolicy(cachedImage))
```

Comment:
Does it pull the image again, even if it exists (or do we need to remove it first)?

Reply:
I don't understand your question fully, but it depends on what `ImagePullPolicy` returns (if it's `true` then it will pull the image, otherwise it skips the pulling). You can think of it like conditionally running `docker pull` before always running `docker run --pull never`.

There are three policies ready to use (in [`PullPolicy`](https://github.com/BenasB/testcontainers-dotnet/blob/feat/601_image_pull_policy/src/Testcontainers/Images/PullPolicy.cs)):

1. `PullPolicy.Never` will always skip the `docker pull` part and the `docker run --pull never` could fail if the cached image is not found (if you didn't run `docker pull` manually). This might be useful in scenarios where you want to manually pull the image before running the tests and not let testcontainers library influence the image in any way.
2. `PullPolicy.Missing` will only do `docker pull` if a cached image is not found. This is what happens by default in CLI `docker run` (no `--pull` argument) when you run for example `docker run alpine:latest` and then a month later (when there's a new `latest` version) run `docker run alpine:latest` again - it will not pull the new `latest` unless you `docker pull` it explicitly/manually. This is also the default pull policy in our library if you don't specify `WithImagePullPolicy` when building the container (meaning no breaking changes done).
3. `PullPolicy.Always` will always do `docker pull`. This is basically running `docker pull` before you run the `docker run --pull never` every time - it will enforce that the newest version of the tag is pulled. If it's up to date, it will return quickly (just like `docker pull` returns quickly when it checks if the image tag is up to date)

Reply:
I understood the policies, but did you check if

```csharp
await this.images.CreateAsync(configuration.Image, authConfig, ct)
  .ConfigureAwait(false);
```

downloads (updates) an image again if it already exists?

Reply:
Ahh sorry, I misunderstood you at first. Yes, [it goes directly](https://github.com/testcontainers/testcontainers-dotnet/blob/develop/src/Testcontainers/Clients/DockerImageOperations.cs#L78) to Docker.DotNet's [`CreateImageAsync`](https://github.com/dotnet/Docker.DotNet/blob/master/src/Docker.DotNet/Endpoints/IImageOperations.cs#L86), which is equivalent to `docker pull`. As I briefly mentioned in #601 as a workaround I could call Docker.Dotnet's `CreateImageAsync` explicitly and achieve what I wanted ("always" policy). 

I will manually test this again just to be sure once I get the opportunity to.

Reply:
> as a workaround I could call Docker.Dotnet's CreateImageAsync explicitly and achieve what I wanted ("always" policy).

Ah sorry, missed that somehow. Do you want to add the tests for `PullPolicy.Never`?

Reply:
Yes I will add the test and mark this PR as ready for review